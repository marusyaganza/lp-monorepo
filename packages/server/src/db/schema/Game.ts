import { Schema, model } from 'mongoose';
import {
  GameConfig,
  Language,
  Game as GameType
} from '../../generated/graphql';

export interface GameConfigType extends GameConfig {
  languages: Language[];
}

const gameSchema = new Schema<GameConfigType>(
  {
    name: { type: String, required: true },
    desc: { type: String, required: true },
    type: {
      type: String,
      required: true,
      enum: GameType,
      immutable: true,
      unique: true
    },
    wordsPerGame: { type: Number, required: true },
    optionsPerGame: { type: Number, required: true },
    minWords: { type: Number, required: true },
    timesToLearn: { type: Number, required: true },
    languages: [{ type: String, enum: Language, required: true }],
    orderNum: { type: Number, unique: true }
  },
  { toObject: { virtuals: true, getters: true, versionKey: false } }
);

gameSchema.virtual('id').get(function () {
  return this._id.toHexString();
});

export const Game = model<GameConfigType>('Game', gameSchema);
