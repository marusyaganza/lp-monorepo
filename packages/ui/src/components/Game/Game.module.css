@import '../../assets/styles/colors.css';
@import '../../assets/styles/font-sizes.css';
@import '../../assets/styles/fonts.css';
@import '../../assets/styles/effects.css';

.gameContainer {
  display: flex;
  flex-direction: column;
  gap: 24px;
  align-items: center;
}

.task {
  text-align: center;
  font-family: var(--font-kalam);
  font-size: var(--font-size-l);
  margin: 0 24px;
  position: relative;
}

.question {
  font-size: var(--font-size-s);
  max-width: 600px;
  display: flex;
  gap: 12px;
  align-items: center;
  margin: 0;
}

.answer {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 40px;
  margin: 0;
}

.container {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin-bottom: 38px;
  gap: 40px;
}

.questionContainer {
  padding: 0 16px;
  display: flex;
  flex-direction: column;
  gap: 12px;
}

.correctAnswer {
  height: 38px;
  box-sizing: border-box;
  padding: 8px;
  color: var(--success-300);
  border: 1px dashed var(--success-200);
  margin: 0;
}

.correctAnswer.hidden {
  opacity: 0;
}

.additionalAudio {
  position: absolute;
  top: 6px;
  right: -28px;
}

.listIcon {
  color: var(--secondary-500);
  flex-shrink: 0;
}

.examplesContainer {
  position: absolute;
  left: 0;
  margin: 30px;
  bottom: 100px;
  max-width: 350px;
  font-family: var(--font-kalam);
}

.examplesHeading {
  font-weight: normal;
  text-align: center;
  margin: 0;
  margin-bottom: 8px;
}

.listIconBook {
  color: var(--primary-100);
  flex-shrink: 0;
}

.example {
  display: flex;
  gap: 12px;
}

.examplesText > i {
  color: var(--primary-500);
  display: block;
}

.additionalInfo {
  position: absolute;
  right: 0;
  margin: 30px;
  max-width: 250px;
  top: 150px;
}

.wordImage {
  width: 200px;
  margin-bottom: 16px;
}
