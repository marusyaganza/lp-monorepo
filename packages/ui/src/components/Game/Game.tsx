import React, {
  FormEventHandler,
  useState,
  useMemo,
  useRef,
  useEffect
} from 'react';
import { cn } from '../../utils/classnames';
import {
  GameQuestionAdditionalInfo,
  Game as GameType
} from '../../generated/graphql';
import { AudioButton } from '../AudioButton/AudioButton';
import { Button } from '../Button/Button';
import { TextInput } from '../TextInput/TextInput';
import styles from './Game.module.css';
import { OptionBox } from '../OptionBox/OptionBox';
import { DictionaryEntity } from '../DictionaryEntity/DictionaryEntity';
import { Icon } from '../Icon/icon';

export interface GameProps {
  task: string;
  question: string[];
  options?: string[];
  type: GameType;
  additionalInfo?: GameQuestionAdditionalInfo | null;
  onSubmit: (value: string) => void;
  /**additional styling */
  className?: string;
  currentResult?: {
    type?: 'initial' | 'success' | 'error';
    correctAnswer?: string;
    incorrectAnswer?: string;
  };
  onNext: () => void;
}
/**Component to display game*/
export const Game = ({
  task,
  question,
  options,
  className,
  type,
  onSubmit,
  currentResult,
  additionalInfo,
  onNext
}: GameProps) => {
  const [value, setValue] = useState('');
  const currentStage = useMemo(
    () => currentResult?.type || 'initial',
    [currentResult?.type]
  );
  const buttonRef = useRef<HTMLButtonElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    // if the component displays the current game result, 'next' button is focused
    if (buttonRef?.current && currentStage !== 'initial') {
      buttonRef.current.focus();
    }
    // The input field should be focused so the user could type the answer
    if (inputRef?.current && currentStage === 'initial') {
      inputRef.current.focus();
    }
  }, [currentStage]);

  const handleChange = (val: string) => {
    setValue(val);
  };

  const handleSubmit: FormEventHandler<HTMLFormElement> = e => {
    e.preventDefault();
    onSubmit(value.trim());
  };

  // Clean the state once the question is answered
  const handleNext = () => {
    setValue('');
    onNext();
  };
  const renderQuestion = () => {
    if (type === GameType.Audio) {
      return (
        <div className={styles.questionContainer}>
          <AudioButton
            autoplay
            iconHeight={50}
            iconWidth={50}
            src={question[0]}
          />
        </div>
      );
    }
    if (question.length === 1) {
      return <p className={styles.question}>{question[0]}</p>;
    }
    return (
      <div className={styles.questionContainer}>
        {question.map(entry => {
          return (
            <p key={entry} className={styles.question}>
              <Icon
                className={styles.listIcon}
                width={20}
                height={20}
                id="comet"
              />
              <DictionaryEntity text={entry} />
            </p>
          );
        })}
      </div>
    );
  };
  const renderAudioButton = () => {
    const hasAudioButton =
      additionalInfo?.audioUrl &&
      (type === GameType.SelectDef ||
        (currentStage !== 'initial' && type !== GameType.Audio));
    return (
      hasAudioButton && (
        <AudioButton
          iconHeight={24}
          iconWidth={24}
          className={styles.additionalAudio}
          src={additionalInfo?.audioUrl || ''}
          autoplay
        />
      )
    );
  };

  const renderInput = () => {
    if (type === GameType.Audio || type === GameType.TypeWord) {
      return (
        <TextInput
          ref={inputRef}
          value={value}
          variant={currentStage}
          name="word"
          onChange={handleChange}
          isDisabled={currentStage !== 'initial'}
        />
      );
    }
    if (options?.length) {
      return (
        <OptionBox
          ref={inputRef}
          value={value}
          isDisabled={currentStage !== 'initial'}
          variant={currentStage}
          correctOption={currentResult?.correctAnswer}
          incorrectOption={currentResult?.incorrectAnswer}
          options={options}
          onChange={handleChange}
        />
      );
    }
    return <p>Game is not found</p>;
  };

  const renderCorrectAnswer = () => {
    const answeredCorrectly =
      !currentResult?.incorrectAnswer || currentStage === 'initial';
    return (
      <span
        className={cn(
          styles.correctAnswer,
          answeredCorrectly ? styles.hidden : ''
        )}
      >
        {!answeredCorrectly && currentResult?.correctAnswer}
      </span>
    );
  };

  return (
    <div className={styles.gameContainer}>
      {!options?.length && renderCorrectAnswer()}
      <article className={cn(styles.container, className)}>
        <p className={styles.task}>
          {task} {renderAudioButton()}
        </p>
        {renderQuestion()}
        <form className={styles.answer} onSubmit={handleSubmit}>
          {renderInput()}
          {currentStage === 'initial' ? (
            <Button disabled={!value} type="submit" variant="secondary">
              Check
            </Button>
          ) : (
            <Button ref={buttonRef} autoFocus onClick={handleNext}>
              Continue
            </Button>
          )}
        </form>
      </article>
    </div>
  );
};
