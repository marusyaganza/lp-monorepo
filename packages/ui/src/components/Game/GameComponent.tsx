import React, { useMemo, useRef, useEffect } from 'react';
import {
  GameQuestionAdditionalInfo,
  Game as GameType,
  Language
} from '../../generated/graphql';
import { Button } from '../Button/Button';
import { DictionaryEntity } from '../DictionaryEntity/DictionaryEntity';
import { Icon } from '../Icon/icon';
import { useModal } from '../Modal/useModal';
import { games } from './games';
import { GameResultType, GameStage, GameState } from '../../types/gameTypes';

import styles from './Game.module.css';
import { FocusableHTMLElement } from '../../types/types';

export interface GameComponentProps {
  task: string;
  question: string[];
  nextQuestion?: string;
  options?: string[] | null;
  type: GameType;
  correctAnswer: string[];
  additionalInfo?: GameQuestionAdditionalInfo | null;
  onSubmit: (value: GameResultType) => void;
  /**additional styling */
  className?: string;
  currentResult?: GameState['currentResult'];
  onNext: () => void;
  language?: Language;
}
/**Component to display game*/
export const GameComponent = ({
  type,
  currentResult,
  additionalInfo,
  onSubmit,
  onNext,
  ...props
}: GameComponentProps) => {
  const currentStage = useMemo(
    () => currentResult?.type || GameStage.Initial,
    [currentResult?.type]
  );
  const showAdditionalInfo = useMemo(
    () => currentStage === 'error' || currentStage === 'success',
    [currentStage]
  );
  const buttonRef = useRef<FocusableHTMLElement>(null);
  const inputRef = useRef<FocusableHTMLElement>(null);
  const { Modal, openModal } = useModal();

  useEffect(() => {
    // if the component displays the current game result, 'next' button is focused
    if (buttonRef?.current && currentStage !== GameStage.Initial) {
      buttonRef.current.focus();
    }
    // The input field should be focused so the user could type the answer
    if (inputRef?.current && currentStage === GameStage.Initial) {
      inputRef.current.focus();
    }
  }, [currentStage]);

  const renderExamples = () => {
    const examples = additionalInfo?.examples;
    if (!examples?.length || !showAdditionalInfo) {
      return;
    }
    return (
      <>
        <h3 className={styles.examplesHeading}>Examples</h3>
        <ul>
          {examples.map(example => {
            const text = example?.text;
            const translation = example?.translation
              ? ` (${example.translation})`
              : '';
            return (
              <li key={text} className={styles.example}>
                <Icon
                  className={styles.listIconBook}
                  width={20}
                  height={20}
                  id="book"
                />
                <DictionaryEntity
                  className={styles.examplesText}
                  text={`${text}${translation}`}
                />
              </li>
            );
          })}
        </ul>
      </>
    );
  };

  const renderAdditionalInfo = () => {
    if (!showAdditionalInfo) {
      return;
    }

    const showButton = !!(
      additionalInfo?.imgUrl || additionalInfo?.examples?.length
    );
    return (
      <section data-cy="additionalInfo" className={styles.additionalInfo}>
        {additionalInfo?.imgUrl && (
          <img
            className={styles.wordImage}
            src={additionalInfo?.imgUrl}
            alt=""
          />
        )}
        {additionalInfo?.shortDef && (
          <DictionaryEntity
            className={styles.examplesText}
            text={additionalInfo?.shortDef}
          />
        )}
        {showButton && (
          <Button
            onClick={openModal}
            variant="tertiary"
            className={styles.additionalInfoBtn}
          >
            Show additional info
          </Button>
        )}
      </section>
    );
  };

  const CurrentGame = games[type];

  return (
    <div className={styles.gameContainer}>
      {renderAdditionalInfo()}
      <CurrentGame
        onSubmit={onSubmit}
        currentStage={currentStage}
        onNext={onNext}
        inputRef={inputRef}
        buttonRef={buttonRef}
        audioUrl={additionalInfo?.audioUrl}
        {...props}
      />
      <section data-cy="examples" className={styles.examplesContainer}>
        {renderExamples()}
      </section>
      <Modal className={styles.modal}>
        <div className={styles.modalContent}>
          {additionalInfo?.imgUrl && (
            <img
              className={styles.modalWordImg}
              src={additionalInfo?.imgUrl}
              alt=""
            />
          )}
          {renderExamples()}
        </div>
      </Modal>
    </div>
  );
};
