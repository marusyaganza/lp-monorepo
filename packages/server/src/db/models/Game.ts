import { GameConfig, Language } from '../../generated/graphql';
import { formatData, formatFilter } from '../helpers';
import { Game } from '../schema/Game';
// import { games as gamesMock } from '../../mocks/games';

export interface GameFilterType extends Partial<GameConfig> {
  languages?: Language;
}

export interface GameModelType {
  findOne: (filter: GameFilterType) => Promise<GameConfig | null>;
  findMany: (filter: GameFilterType) => Promise<GameConfig[] | null>;
}

export const GameModel: GameModelType = {
  async findOne(filter) {
    const game = await Game.findOne(formatFilter(filter));
    return formatData(game);
  },

  async findMany(filter) {
    const games = await Game.find(formatFilter(filter));
    // await Game.insertMany(gamesMock);
    return games;
  }
};
