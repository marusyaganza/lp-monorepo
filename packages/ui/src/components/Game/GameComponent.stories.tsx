import React from 'react';
import { GameComponent, GameComponentProps } from './GameComponent';
import '../../assets/styles/common-styles.css';
import { Game as GameType } from '../../generated/graphql';
import { styledPreviewDecorator } from '../../storybook-decorators';
import type { Meta } from '@storybook/react';
import { GameStage } from '../../types/gameTypes';

const meta: Meta<typeof GameComponent> = {
  title: 'game/GameComponent',
  component: GameComponent,
  decorators: [styledPreviewDecorator()]
};

const games = [
  {
    task: "Type the word that you've heard",
    type: GameType.Audio,
    correctAnswer: ['wheel'],
    question: [
      'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    ]
  },
  {
    task: 'Type a word that means',
    type: GameType.TypeWord,
    correctAnswer: ['wheel'],
    question: [
      'a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle',
      'a contrivance or apparatus having as its principal part a wheel: such as',
      'a chiefly medieval instrument of torture designed for mutilating a victim (as by stretching or disjointing)'
    ],
    additionalInfo: {
      audioUrl:
        'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    }
  },
  {
    task: 'Select a word that means',
    type: GameType.SelectWord,
    question: [
      'a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle',
      'a contrivance or apparatus having as its principal part a wheel: such as',
      'a chiefly medieval instrument of torture designed for mutilating a victim (as by stretching or disjointing)'
    ],
    options: ['forefront', 'name', 'apple', 'wheel', 'pear', 'steep'],
    correctAnswer: ['wheel'],
    additionalInfo: {
      audioUrl:
        'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    }
  },
  {
    task: 'Select a definition that means',
    type: GameType.SelectDef,
    question: ['wheel'],
    correctAnswer: [
      'a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle'
    ],
    options: [
      'a person or thing that is different from or in a position away from others in the group',
      'a small piece of rubber or other material that is used to erase something you have written or drawn —called also (British) rubber',
      'to get information or a reaction from somebody, often with difficulty.',
      'a brief authoritative formula of religious belief',
      'a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle',
      'a cat or kitten : pussycat—used especially by children or when talking to children'
    ],
    additionalInfo: {
      audioUrl:
        'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    }
  },
  {
    task: 'Conjugate the verb',
    type: GameType.Conjugation,
    question: ['caer'],
    correctAnswer: ['caigo', 'caes', 'cae', 'caemos', 'ca\u00e9is', 'caen'],
    additionalInfo: {
      audioUrl:
        'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    }
  },
  {
    question: ['ecologista'],
    type: GameType.Gender,
    correctAnswer: ['masculine', 'feminine'],
    additionalInfo: {
      audioUrl:
        'https://voice.reverso.net/RestPronunciation.svc/v1/output=json/GetVoiceStream/voiceName=Maria22k?voiceSpeed=80&inputText=ZWNvbG9naXN0YQ==',
      shortDef: '<b>ecologista means</b> ecologist, environmentalist'
    }
  }
];

export const GameDefault = (args: GameComponentProps) =>
  games.map(game => {
    return (
      <div key={game.type} className="presentationBox" style={{ width: 600 }}>
        <GameComponent {...args} {...game} />
      </div>
    );
  });

export const GameSuccess = (args: GameComponentProps) =>
  games.map(game => {
    return (
      <div key={game.type} className="presentationBox" style={{ width: 600 }}>
        <GameComponent
          {...args}
          {...game}
          currentResult={{ type: GameStage.Success }}
        />
      </div>
    );
  });

export const GameError = (args: GameComponentProps) =>
  games.map(game => {
    return (
      <div key={game.type} className="presentationBox" style={{ width: 600 }}>
        <GameComponent
          {...args}
          {...game}
          currentResult={{
            type: GameStage.Error
          }}
        />
      </div>
    );
  });

export default meta;
