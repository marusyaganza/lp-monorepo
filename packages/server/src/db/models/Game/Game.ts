import {
  GameConfig,
  Language,
  Game as GameType
} from '../../../generated/graphql';
import { Game } from '../../schema/Game';

export interface GameFilterType {
  languages?: Language;
  type?: GameType;
}

export interface GameModelType {
  findOne: (filter: GameFilterType) => Promise<GameConfig | null | undefined>;
  findMany: (
    filter: GameFilterType
  ) => Promise<GameConfig[] | null | undefined>;
}

export const GameModel: GameModelType = {
  async findOne(filter) {
    const game = await Game.findOne(filter);
    return game?.toObject();
  },

  async findMany(filter) {
    const games = await Game.find(filter).sort('orderNum').exec();
    return games;
  }
};
