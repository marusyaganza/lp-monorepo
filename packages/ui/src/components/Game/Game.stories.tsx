import React from 'react';
import { Game, GameProps } from './Game';
import '../../assets/styles/common-styles.css';
import { Game as GameType } from '../../generated/graphql';
import { styledPreviewDecorator } from '../../storybook-decorators';
import type { Meta } from '@storybook/react';

const meta: Meta<typeof Game> = {
  title: 'Game',
  component: Game,
  decorators: [styledPreviewDecorator()]
};

const games = [
  {
    task: "Type the word that you've heard",
    type: GameType.Audio,
    question: [
      'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    ]
  },
  {
    task: 'Type a word that means',
    type: GameType.TypeWord,
    question: [
      'a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle',
      'a contrivance or apparatus having as its principal part a wheel: such as',
      'a chiefly medieval instrument of torture designed for mutilating a victim (as by stretching or disjointing)'
    ],
    additionalInfo: {
      audioUrl:
        'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    }
  },
  {
    task: 'Select a word that means',
    type: GameType.SelectWord,
    question: [
      'a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle',
      'a contrivance or apparatus having as its principal part a wheel: such as',
      'a chiefly medieval instrument of torture designed for mutilating a victim (as by stretching or disjointing)'
    ],
    options: ['forefront', 'name', 'apple', 'wheel', 'pear', 'steep'],
    additionalInfo: {
      audioUrl:
        'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    }
  },
  {
    task: 'Select a definition that means',
    type: GameType.SelectDef,
    question: ['wheel'],
    options: [
      'a person or thing that is different from or in a position away from others in the group',
      'a small piece of rubber or other material that is used to erase something you have written or drawn —called also (British) rubber',
      'to get information or a reaction from somebody, often with difficulty.',
      'a brief authoritative formula of religious belief',
      'a circular frame of hard material that may be solid, partly solid, or spoked and that is capable of turning on an axle',
      'a cat or kitten : pussycat—used especially by children or when talking to children'
    ],
    additionalInfo: {
      audioUrl:
        'https://media.merriam-webster.com/audio/prons/en/us/mp3/w/wheel001.mp3'
    }
  }
];

export const GameDefault = (args: GameProps) =>
  games.map(game => {
    return (
      <div key={game.type} className="presentationBox" style={{ width: 600 }}>
        <Game {...args} {...game} />
      </div>
    );
  });

export const GameSuccess = (args: GameProps) =>
  games.map(game => {
    return (
      <div key={game.type} className="presentationBox" style={{ width: 600 }}>
        <Game
          {...args}
          {...game}
          currentResult={{ type: 'success', correctAnswer: 'wheel' }}
        />
      </div>
    );
  });

export const GameError = (args: GameProps) =>
  games.map(game => {
    return (
      <div key={game.type} className="presentationBox" style={{ width: 600 }}>
        <Game
          {...args}
          {...game}
          currentResult={{
            type: 'error',
            correctAnswer: 'wheel',
            incorrectAnswer: 'flower'
          }}
        />
      </div>
    );
  });

export default meta;
