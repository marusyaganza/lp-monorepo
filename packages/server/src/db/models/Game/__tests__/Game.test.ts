import { GameModel } from '../Game';
import { Game, Language } from '../../../../generated/graphql';
import {
  connectToDb,
  disconnectFromDb,
  dropDb,
  seedDb,
  baseSnapshotConfig
} from '../../../../tests/helpers';

describe('Game findMany', () => {
  beforeEach(async () => {
    await connectToDb();
    await dropDb();
    await seedDb();
  });

  afterEach(async () => {
    await dropDb();
    await disconnectFromDb();
  });

  test('English', async () => {
    const result = await GameModel.findMany({
      languages: Language.English
    });
    expect(result?.length).toBe(4);
    expect(result?.map(item => item?.type)).toEqual([
      'AUDIO',
      'SELECT_DEF',
      'SELECT_WORD',
      'TYPE_WORD'
    ]);
  });
  test('Spanish', async () => {
    const result = await GameModel.findMany({
      languages: Language.Spanish
    });

    expect(result?.length).toBe(6);
    expect(result?.map(item => item?.type)).toEqual([
      'AUDIO',
      'SELECT_DEF',
      'SELECT_WORD',
      'TYPE_WORD',
      'CONJUGATION',
      'GENDER'
    ]);
  });
});

const gameTypes = Object.values(Game);

describe('Game findOne', () => {
  beforeEach(async () => {
    await connectToDb();
    await dropDb();
    await seedDb();
  });

  afterEach(async () => {
    await dropDb();
    await disconnectFromDb();
  });
  gameTypes.forEach(game => {
    test(`with ${game} game, English`, async () => {
      const result = await GameModel.findOne({
        type: game
      });
      expect(result).toMatchSnapshot(baseSnapshotConfig);
    });
  });
});
