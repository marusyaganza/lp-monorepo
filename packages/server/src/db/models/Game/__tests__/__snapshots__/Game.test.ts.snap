// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game findOne with AUDIO game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Listen, Type, Master. Sharpen language skills through auditory engagement.",
  "id": Any<String>,
  "languages": [
    "ENGLISH",
    "SPANISH",
  ],
  "minWords": 1,
  "name": "WordEcho",
  "optionsPerGame": 0,
  "orderNum": 0,
  "timesToLearn": 5,
  "type": "AUDIO",
  "wordsPerGame": 6,
}
`;

exports[`Game findOne with CONJUGATION game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Master verb conjugations across tenses and languages in fast-paced, interactive challenges!",
  "id": Any<String>,
  "languages": [
    "SPANISH",
  ],
  "minWords": 1,
  "name": "Conjugation Quest",
  "optionsPerGame": 0,
  "orderNum": 4,
  "timesToLearn": 5,
  "type": "CONJUGATION",
  "wordsPerGame": 2,
}
`;

exports[`Game findOne with GENDER game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Choose nouns and match their gender. Test linguistic skills in thrilling rounds!",
  "id": Any<String>,
  "languages": [
    "SPANISH",
  ],
  "minWords": 1,
  "name": "GenderWord Challenge",
  "optionsPerGame": 0,
  "orderNum": 5,
  "timesToLearn": 5,
  "type": "GENDER",
  "wordsPerGame": 5,
}
`;

exports[`Game findOne with SELECT_DEF game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Define & Conquer.Match words with their meanings intuitively.",
  "id": Any<String>,
  "languages": [
    "ENGLISH",
    "SPANISH",
  ],
  "minWords": 1,
  "name": "LexiSense",
  "optionsPerGame": 6,
  "orderNum": 1,
  "timesToLearn": 5,
  "type": "SELECT_DEF",
  "wordsPerGame": 6,
}
`;

exports[`Game findOne with SELECT_WORD game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Associate definitions with the right words. A captivating word puzzle adventure!",
  "id": Any<String>,
  "languages": [
    "ENGLISH",
    "SPANISH",
  ],
  "minWords": 1,
  "name": "Definify: Match & Learn",
  "optionsPerGame": 6,
  "orderNum": 2,
  "timesToLearn": 5,
  "type": "SELECT_WORD",
  "wordsPerGame": 6,
}
`;

exports[`Game findOne with TYPE_WORD game, English 1`] = `
{
  "_id": Any<Object>,
  "desc": "Type words matching given definitions. Boost vocabulary in engaging rounds!",
  "id": Any<String>,
  "languages": [
    "ENGLISH",
    "SPANISH",
  ],
  "minWords": 1,
  "name": "WordRecall: Define & Type",
  "optionsPerGame": 0,
  "orderNum": 3,
  "timesToLearn": 5,
  "type": "TYPE_WORD",
  "wordsPerGame": 6,
}
`;
